File: Toolboxes/Tisean_3.0.1/unsupported/surrogates.c
/* Create Surrogate data */

/****************************************************************/
/* this source is freely distributable                          */
/* copyright (C) Andreas Schmitz (1997)                         */
/* Wuppertal University, Germany                                */
/****************************************************************/

#include <math.h>
#include <stdlib.h>
#include <stddef.h>
#include <stdio.h>
#include <string.h>
#include <time.h>
#include <errno.h>
#include "arguments.h"
#include "surro.h"
#include "dft.h"
#include "rank.h"
#include "rand.h"

#define MAXNAMELEN 512      /* maximal lenght of file name */
#define DEF_POINTS 10000000 /* default number of points */
#define DEF_ITER   100000   /* default iterations for polishing */
#define DEF_NUMB   1        /* default number of surrogates */

void usage(char *progname)
{
   fprintf(stderr,"\nUsage: %s [-n# -i# -I# -l# -x# -c# -h] file\n",progname);
   fprintf(stderr,"\t-n <number of surrogates (1)>\n");
   fprintf(stderr,"\t-i <number of iterations (until no change)>\n");
   fprintf(stderr,"\t-I <seed for random numbers (0 = call to time())>\n");
   fprintf(stderr,"\t-l <number of points (whole file)>\n");
   fprintf(stderr,"\t-x <number of values to be skipped (0)>\n");
   fprintf(stderr,"\t-c <column to be read (1 or file,#)>\n");
   fprintf(stderr,"\t-h <show this message>\n\n");
   exit(0);
}

int main(int argc,char *argv[])
{
   double *y=NULL, *z=NULL, *ampl_org, *arr_sort, sy, sy2, syz;
   char file[MAXNAMELEN], *outfile; 
   long number, col, iter, it, nmax, nexcl, n, i, seed=67654748, filen, iget;
   FILE *fpin=stdin, *fpout=stdout;

   whatido(argc,argv,"create iterative surrogate data",argv[0]);
   nmax=lcan(argc,argv,'l',DEF_POINTS);
   nexcl=lcan(argc,argv,'x',0);
   col=lcan(argc,argv,'c',0);
   number=lcan(argc,argv,'n',DEF_NUMB);
   iter=lcan(argc,argv,'i',DEF_ITER);
   seed=-lcan(argc,argv,'I',seed);
   if(seed==0) seed=time(NULL);
   rnd_init((unsigned long)seed);
   outfile=getout(argc,argv,&iget);

   col=getfile(file,nthstring(argc,argv,1),col);
   if(file[0]=='-')
      file[0]='\0';
   else {
      fprintf(stderr,"trying to open file %s\n", file);
      fpin=fopen(file,"r");
      if(fpin==NULL) {
	fprintf(stderr,"cannot open file %s\n",file);
	return 5;
      }
   }
   if(number>1) iget=1; 
   if(iget){
      if(file[0]=='\0') strcpy(file,"stdin");
      putcol(file,col);
      strcat(file, "_surr");
   }
   else
      file[0]='\0';

   if(outfile[0]!='\0') strcpy(file,outfile);
   filen=(long)strlen(file);

   for(n=0; n<nexcl; n++) if(readdum(fpin)==EOF) return 4;
   for(n=0; n<nmax; n++){
      y=(double *)realloc(y, (size_t)((n+1)*sizeof(double)));
      if(readval(fpin,y+n,col)==EOF) break;
   }
   fprintf(stderr,"read %ld values\n",n);

   nmax=strip_primes(n);   /* reduce filelength -> low prime factors */
   if (n!=nmax) fprintf(stderr,"shortened to length: %ld\n",nmax);

   arr_sort=(double *)malloc((size_t)(nmax*sizeof(double)));
   ampl_org=(double *)malloc((size_t)((nmax/2+1)*sizeof(double)));
   z=(double *)malloc(nmax*sizeof(double));
   if (arr_sort==NULL || ampl_org==NULL || z==NULL) {
      fprintf(stderr,"no memory\n");
      exit(5);
   }

   dft(y,nmax,ampl_org,NULL); /* original amplitudes */

   for (i=0; i<nmax; i++)    /* make a sorted copy of the org data */
      arr_sort[i]=y[i];   
   sort(nmax,arr_sort);
      
   for (i=0; i<number; i++) {
      if(file[0]!='\0'){
         if(number>1) sprintf(file+filen,"_%3.3ld",i+1);
         fprintf(stderr,"%s ",file);
         fpout=fopen(file,"w");
         if(fpout==NULL) {
	   fprintf(stderr,"cannot open file %s\n",file);
	   return 5;
         }
      }
      it=pol_surr(nmax,y,z,arr_sort,ampl_org,iter);
      for(n=0, sy=0, sy2=0, syz=0; n<nmax; n++){
	 sy+=y[n];
         sy2+=y[n]*y[n];
         syz+=(y[n]-z[n])*(y[n]-z[n]);
      }
      sy=sqrt((sy2/nmax)-(sy/nmax)*(sy/nmax));
      syz=sqrt(syz/nmax);
      fprintf(stderr,"(%ld iterations, relative discrepancy  %lf)\n", 
         it,syz/sy);
      for(n=0; n<nmax; n++)
         fprintf(fpout,"%lf %lf\n",y[n],z[n]);
      fclose(fpout);
   }
   
   free(y);
   free(arr_sort);

   return 0;
}

File: Toolboxes/Tisean_3.0.1/unsupported/surrogates.f
c Create surrogate data
c Copyright (C) T. Schreiber (1999)

      parameter(nx=1000000)
      dimension xx(nx), x(nx), y(nx), xamp(nx), xsort(nx), list(nx)
      character*72 file, fout
      data nsur/1/, imax/-1/
      external rand

      call whatido("Create Surrogate data")
      nmax=ican("l",nx)
      nexcl=ican("x",0)
      jcol=ican("c",0)
      nsur=min(999,ican("n",nsur))
      imax=ican("i",imax)
      ispec=ican("S",0)
      r=rand(sqrt(fcan("I",0.0)))
      iverb=igetv(15)
      isout=igetout(fout,iexv(iverb,1))

      call nthstring(1,file)
      call readfile(nmax,xx,nexcl,jcol,file,iexv(iverb,1))
      nmaxp=nless(nmax)
      if(nmaxp.ne.nmax.and.iexv(iverb,1).ne.0) 
     .   write(istderr(),*) "surrogates: using first ", nmaxp
      if(file.eq."-") file="stdin"
      if(isout.eq.1) call addsuff(fout,file,"_surr")
      if(nsur.gt.1.and.isout.eq.1) call suffix(fout,"_000")

      do 10 isur=1,nsur
         if(nsur.gt.1.and.isout.eq.1) 
     .      write(fout(index(fout," ")-3:72),'(i3.3)') isur
         do 20 n=1,nmaxp
            x(n)=xx(n)
            y(n)=x(n)
            xamp(n)=x(n)
 20         xsort(n)=x(n)
         do 30 n=1,nmaxp
            nper=min(int(rand(0.0)*nmaxp)+1,nmaxp)
            h=x(n)
            x(n)=x(nper)
 30         x(nper)=h
         call store_spec(nmaxp,xamp,0)
         call sort(nmaxp,xsort,list)
         it=-1
         dspec=r1mach(2)
 1       it=it+1
         do 40 n=1,nmaxp
 40         y(n)=x(n)
         ds0=dspec
         dspec=totospec(nmaxp,xamp,y)
         if(imax.ge.0.and.it.ge.imax) goto 2
         call todist(nmaxp,xsort,y,x)
         if(dspec.lt.ds0) goto 1
 2       continue
         if(ispec.gt.0) then
            call writefile(nmaxp,y,fout,iexv(iverb,1))
         else
            call writefile(nmaxp,x,fout,iexv(iverb,1))
         endif
 10      if(iexv(iverb,2).ne.0) write(istderr(),*) 
     .      fout(1:index(fout," ")), ' (', it, 
     .      ' iterations, relative discrepancy ', dspec,   ')'
      end

      subroutine usage()
c usage message

      call whatineed(
     .   "[-n# -i# -S -I# -o outfile -l# -x# -c# -V# -h] file")
      call popt("n","number of surrogates (1)")
      call popt("i","number of iterations (until no change)")
      call popt("S","make spectrum exact rather than distribution")
      call popt("I","seed for random numbers")
      call popt("l","number of values to be read (all)")
      call popt("x","number of values to be skipped (0)")
      call popt("c","column to be read (1 or file,#)")
      call pout("file_surr(_nnn)")
      call pall()
      stop
      end

      function totospec(nmax,a,x)
      parameter(nx=1000000)
      dimension x(nmax), a(nmax), w(nx), w1(nx), w2(nx), iw(15)

      if(nmax.gt.nx) stop "totospec: make nx larger."
      do 10 n=1,nmax
 10      w(n)=x(n)
      call rffti1(nmax,w2,iw)  
      call rfftf1(nmax,x,w1,w2,iw)
      do 20 n=1,nmax
 20      x(n)=x(n)/real(nmax)
      x(1)=sqrt(a(1))
      do 30 n=2,(nmax+1)/2
         ab=a(2*n-2)/(x(2*n-2)**2+x(2*n-1)**2)
         x(2*n-2)=x(2*n-2)*sqrt(ab)
 30      x(2*n-1)=x(2*n-1)*sqrt(ab)
      if(mod(nmax,2).eq.0) x(nmax)=sqrt(a(nmax))
      call rfftb1(nmax,x,w1,w2,iw)
      totospec=0
      do 40 n=1,nmax
 40      totospec=totospec+(x(n)-w(n))**2
      totospec=sqrt(totospec/nmax)
      end


